KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,238 @@
+
+#### Finalize Contract
+
+Once the re-distribution a.k.a rebalance is executed by the Core, the status of the smart contract will be finalized and any modifications to the storage data will be restricted.
+
+## Backwards Compatibility
+
+<!-- All KIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The KIP must explain how the author proposes to deal with these incompatibilities. KIP submissions without a sufficient backwards compatibility treatise may be rejected outright. The authors should answer the question: "Does this KIP require a hard fork?" -->
+
+- The foundation should deploy new TreasuryRebalance contract to record the token redistribution.
+- To rebalance the funds and redistribute in a consistent manner the foundation should burn the designated funds before re-distribution.
+- This does not affect the backward compatibility as this a newly dpeloyed contract
+
+## Implementation
+
+<!--The implementations must be completed before any KIP is given status "Final", but it need not be completed before the KIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
+
+#### Example implementation
+
+```solidity
+interface ITreasuryRebalance {
+    // Events
+    event ContractDeployed(
+        uint256 rebalanceBlockNumber,
+        uint256 deployedBlockNumber
+    );
+    event RetiredRegistered(address retired);
+    event RetiredRemoved(address retired, uint256 retiredCount);
+    event NewbieRegistered(address newbie, uint256 fundAllocation);
+    event NewbieRemoved(address newbie, uint256 newbieCount);
+    event Approved(address retired, address approver, uint256 approversCount);
+    event StatusChanged(Status status);
+    event Finalized(string memo);
+
+    // Enums
+    enum Status {
+        Initialized,
+        Registered,
+        Approved,
+        Finalized
+    }
+
+    struct Retired {
+        address retired;
+        address[] approvers;
+    }
+
+    struct Newbie {
+        address newbie;
+        uint256 amount;
+    }
+
+    // State variables
+    function status() external view returns (Status);
+    function rebalanceBlockNumber() external view returns (uint256);
+    function memo() external view returns (string memory);
+
+    // State changing functions
+    function registerRetired(address retiredAddress) external;
+    function removeRetired(address retiredAddress) external;
+    function registerNewbie(address newbieAddress, uint256 _amount) external;
+    function removeNewbie(address newbieAddress) external;
+    function approve(address retiredAddress) external;
+    function finalizeRegistration() external;
```

Comment:
Who can call the finalize functions?
if we cannot specify that information using modifier, we need to add a comment for the info.

Reply:
Sure will add comments for the interface